src/functions.php

```diff
@@ -25,15 +25,29 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
             $loop->cancelTimer($timer);
             $reject($v);
         });
+    }, function ($resolveUnused, $reject) use ($loop, &$timer, $promise) {
+        // cancelling this promise will cancel the timer and reject
+        $loop->cancelTimer($timer);
+        $reject(new \RuntimeException('Timeout cancelled'));
```

Comment:
Not sure how to gist this. We should simply do not take responsibility for the cancelling behaviour (which is not necessarily rejection) from the root promise. Why reject here if the root promise decides to do something else? We should only recject if an actual timeout occurs.